verilog/CST/type_test.cc

```diff
@@ -407,6 +407,154 @@ TEST(GetVariableDeclaration, FindPackedDimensionFromDataDeclaration) {
+      {""},
+      {"module m(logic x);\nendmodule"},
+      {"module m(", {kTag, "bus"}, " x);\nendmodule"},
+  };
+  for (const auto& test : kTestCases) {
+    TestVerilogSyntaxRangeMatches(
+        __FUNCTION__, test, [](const TextStructureView& text_structure) {
+          const auto& root = text_structure.SyntaxTree();
+          const auto& types =
+              FindAllDataTypeDeclarations(*ABSL_DIE_IF_NULL(root));
+
+          std::vector<TreeSearchMatch> names;
+          for (const auto& decl : types) {
+            const auto* name = GetTypeIdentifierFromDataType(*decl.match);
+            if (name == nullptr) {
+              continue;
+            }
+            names.emplace_back(TreeSearchMatch{name, {/* ignored context */}});
+          }
+          return names;
+        });
+  }
+}
+
+TEST(GetDataImplicitIdDimensions, GetTypeOfDataImplicitIdDimensions) {
+  constexpr int kTag = 1;  // value doesn't matter
+  const SyntaxTreeSearchTestCase kTestCases[] = {
+      {""},
+      {"module m(logic x);\nendmodule"},
+      {"struct {struct {int x;} var2;} var1;"},
+      {"struct {", {kTag, "my_type"}, " var2;} var1;"},
+      {"union {union {int x;} var2;} var1;"},
+      {"union {", {kTag, "my_type"}, " var2;} var1;"},
+  };
+  for (const auto& test : kTestCases) {
+    TestVerilogSyntaxRangeMatches(
+        __FUNCTION__, test, [](const TextStructureView& text_structure) {
+          const auto& root = text_structure.SyntaxTree();
+          const auto& types =
+              FindAllDataTypeImplicitIdDimensions(*ABSL_DIE_IF_NULL(root));
+
+          std::vector<TreeSearchMatch> inner_types;
+          for (const auto& decl : types) {
+            const auto* inner_type =
+                GetTypeOfDataTypeImplicitIdDimensions(*decl.match);
+            if (inner_type == nullptr) {
+              continue;
+            }
+            inner_types.emplace_back(
+                TreeSearchMatch{inner_type, {/* ignored context */}});
+          }
+          return inner_types;
+        });
+  }
+}
+
+TEST(GetDataImplicitIdDimensions, GetNameOfDataImplicitIdDimensions) {
+  constexpr int kTag = 1;  // value doesn't matter
+  const SyntaxTreeSearchTestCase kTestCases[] = {
+      {""},
+      {"module m(logic x);\nendmodule"},
+      {"struct {int ", {kTag, "xx"}, ";} var1;"},
+      {"union {int ", {kTag, "xx"}, ";} var1;"},
+      {"struct {int ",
+       {kTag, "xx"},
+       ";\n struct {int  ",
+       {kTag, "xx"},
+       ";} ",
+       {kTag, "var2"},
+       ";} var1;"},
+      {"union {int ",
+       {kTag, "xx"},
+       ";\n struct {int  ",
+       {kTag, "xx"},
+       ";} ",
+       {kTag, "var2"},
+       ";} var1;"},
+      {"union {int ", {kTag, "xx"}, ";} var1;"},
+      {"typedef union {int ", {kTag, "xx"}, ";} var1;"},
+      {"typedef struct {int ", {kTag, "xx"}, ";} var1;"},
+      {"struct {some_type ", {kTag, "xx"}, ";} var1;"},
+      {"union {some_type ", {kTag, "xx"}, ";} var1;"},
+  };
```

Comment:
Suggest at least one test case with multiple struct members.

* `struct { int xx; int yy }`
* `struct { int xx, yy }`

Does this work with nested structs?

```systemverilog
typedef struct {
  struct {
    int xx;
  } bar;
} foo_t;
```

Reply:
The CST for:
```
struct { int xx, yy; } var1;
```

```
Node @0 (tag: kStructUnionMember) {
                    Node @1 (tag: kDataTypeImplicitIdDimensions) {
                      Node @0 (tag: kDataType) {
                        Node @0 (tag: kDataTypePrimitive) {
                          Leaf @0 (#"int" @11-14: "int")
                          Node @2 (tag: kPackedDimensions) {
                          }
                        }
                      }
                      Leaf @2 (#SymbolIdentifier @15-17: "xx")
                      Node @3 (tag: kUnpackedDimensions) {
                      }
                    }
                    Leaf @3 (#',' @18-19: ",")
                    Node @4 (tag: kVariableDeclarationAssignmentList) {
                      Node @0 (tag: kVariableDeclarationAssignment) {
                        Leaf @0 (#SymbolIdentifier @20-21: "y")
                        Node @1 (tag: kUnpackedDimensions) {
                        }
                      }
                    }
                    Leaf @5 (#';' @21-22: ";")
                  }
                }
```

Is the node for `yy` correct?

Reply:
What should probably happen is that the CST be reshaped so that `xx` becomes part of the `kVariableDeclarationAssignmentList`.  If left as-is, would you be able to make this work?  If not, file a bug for me with the above example.

Reply:
This works for the extractor. no change need for now.